### ISSUE TITLE
subscribe onDragStart & onDragEnd

### ISSUE THREAD

**Issue by dgtllion** [minop1205/react-dnd-treeview#84]
Hi @minop1205,

Is there a way how i can subscribe on those events of ul/li?
I have my `CustomNode.tsx`, but if i add `onDragStart` and `onDragEnd` then it doesnt do anything. 

Or may be can you pass in `Node.tsx` the ref?

```
    const params = {
        depth: props.depth,
        isOpen: open,
        hasChild,
        onToggle: handleToggle,
        ref: ref
    };
```
Then i can catch the ref in my `CustomNode.tsx`.

**Comment by minop1205** [comment]
@dgtllion In fact, it is the li element that wraps the CustomNode that is being dragged, so setting an event handler on the CustomNode itself will not fire the event.

Therefore, we may need to add a new API to the Tree component.

**Comment by dgtllion** [comment]
Ok, there will be a new release for it? What and when to expect?
But the second question is it possible to pass the `ref` with params?

**Comment by dgtllion** [comment]
I made a workaround which is working, may be its good to add to examples or update documentation? (You can also pass Components to `listItemComponent={Component}`)

i created `MainNode.js` looks like this:
```
const MainNode = forwardRef((props, ref) => {
  const { children, role, className } = props;
  // Here we get current node
  const { node } = children[0].props;

  const onDragStart = (ev) => {
    //Do something
  };

  const onDragEnd = (ev) => {
    //Do something 
  };
  return (
    <li
      ref={ref}
      onDragStart={onDragStart}
      onDragEnd={onDragEnd}
      className={className}
      role={role}
    >
      {props.children}
    </li>
  );
});
```
And this `MainNode` i passed to my `Tree`

```
import MainNode from './MainNode';
<Tree
 tree={treeData}
 listItemComponent={MainNode}
...
/>
```

**Comment by minop1205** [comment]
@dgtllion Thank you!

I tried the code you provided. The onDragStart/onDragEnd does indeed fire.
(However, if I repeatedly drag/drop a node, the event fires multiple times, so I may need to disable the listener somewhere.)

However, in react-dnd-treeview, the CustomNode wrapper element (`<li>`) is under the control of **react-dnd**, and if the event handler is not registered via react-dnd, it may cause unexpected behavior.

In addition, this approach requires a slightly more complex implementation on the main application side, so I think it would be easier to add the API to the TreeView component.

The implementation of the API itself is simple, so we hope to release a version (v1.6.0) that includes this modification by the end of this week.

**Comment by dgtllion** [comment]
Thank you... I will wait for it.

**Comment by minop1205** [comment]
@dgtllion I have just made an alpha release of v1.6.0.
Please install and try it with the following command.

```
npm install @minoru/react-dnd-treeview@alpha
```

In this version, two APIs, `onDragStart` and `onDragEnd`, have been added. Please use them as follows.

```
    <Tree
      {...props}
      onDrop={handleDrop}
      onDragStart={(node) => {
        //do something
      }}
      onDragEnd={(node) => {
        //do something
      }}
    />
```

In both handlers, the node to be dragged is passed.

When a drag operation is canceled (by pressing the escape key or dropping into a non-dropable area), only onDragEnd will fire.
If dropped, onDrop and onDragEnd will fire in that order.

If there are no problems with this version, we will update the documentation and officially release v1.6.0.

**Comment by dgtllion** [comment]
@minop1205 works good! Now the only thing that i miss is `event` itself. Can you also include it to params?
In my case its necessary to have event.

```
onDragStart={(node, event) => {}}
onDragEnd={(node, event) => {}}
```

And may be its also a good idea to pass the events to other functions where events are fired? like `onDrop`?

**Comment by minop1205** [comment]
@dgtllion Thank you.

I looked into the interface of the dependent package, `react-dnd`, but I couldn't find a way to access the native events at drag start and drag end.
https://react-dnd.github.io/react-dnd/docs/api/use-drag

So, if you need to access native events, it might be better to implement it in the following way that you first suggested.
https://github.com/minop1205/react-dnd-treeview/issues/84#issuecomment-972592697

If you don't mind me asking, could you tell me why you need to access native events?

**Comment by dgtllion** [comment]
@minop1205 i need to set some text which will be dropped to WYSIWYG editor. ;)
```
event.dataTransfer.setData('text/plain, 'Some text');
```

Strange that `react-dnd` doesn't provide it... Then the only way now is to do it like i told.

I think the documentation needs to be updated. `listItemComponent: Component | string`

And btw your JSX examples in documentation doesn't work as well.

**Comment by minop1205** [comment]
@dgtllion In the end, I thought the simplest way was to add a ref to the first render option that you suggested.

```
            render={(
              node: NodeModel,
              { depth, isOpen, onToggle, containerRef }
            ) => (
              <CustomNode
                node={node}
                depth={depth}
                isOpen={isOpen}
                onToggle={onToggle}
                containerRef={containerRef}
              />
            )}

```

The containerRef above contains a DOM reference to the container (li element) of the custom node.
You can then access the native drag event by adding an event listener on the custom node side as shown below.

```
const handleDragStart = (e: DragEvent) => {
  // exclude event bubbling
  if (e.target === e.currentTarget) {
    console.log(`drag start: ${props.node.id}`);
  }
};

const handleDragEnd = (e: DragEvent) => {
  // exclude event bubbling
  if (e.target === e.currentTarget) {
    console.log(`drag end ${props.node.id}`);
  }
};

useEffect(() => {
  props.containerRef.current.addEventListener("dragstart", handleDragStart);
  props.containerRef.current.addEventListener("dragend", handleDragEnd);

  return () => {
    props.containerRef.current.removeEventListner("dragstart", handleDragStart);
    props.containerRef.current.removeEventListner("dragend", handleDragEnd);
  };
}, []);
```

Here is a sample that actually works.
https://codesandbox.io/s/issue84-zfd57?file=/src/CustomNode.tsx

I have just released this package as an alpha version, so please try it out.
Please let me know if this version seems to solve the problem.

```
npm install @minoru/react-dnd-treeview@alpha
```

> And btw your JSX examples in documentation doesn't work as well.

Also, what part of the document is this referring to? Please let me know so I can fix it.

**Comment by dgtllion** [comment]
@minop1205 this works nice! thank you!

> Also, what part of the document is this referring to? Please let me know so I can fix it.

Now i see it works... strange. nvm :)

**Comment by dgtllion** [comment]
When do you expect to make a release 1.6.0?

**Comment by minop1205** [comment]
@dgtllion Did you solve the problem with this version?
https://github.com/minop1205/react-dnd-treeview/issues/84#issuecomment-978713803

If there are no problems, i will add more tests and fix the documentation after this, so i plan to release it by the end of next week at the latest.

**Comment by dgtllion** [comment]
@minop1205 no issues.. Everything works ;)
Ok thanks!

**Comment by minop1205** [comment]
Ihave just released v1.6.0.
See here for the release details.

https://github.com/minop1205/react-dnd-treeview/blob/next/CHANGELOG.md

Thank you
